samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -110,17 +109,92 @@ public ZkKeyBuilder getKeyBuilder() {
+    }
+    return ephemeralPath;
+  }
+
+  /**
+   * Registers a processor with zookeeper. Creates a sequential ephemeral node in zookeeper(which will be deleted
+   * on zkClient session timeout) for the processor under processors directory.
+   * @param processorData to register in zookeeper.
+   * @return the sequential ephemeral path created in zookeeper.
+   */
+  String registerProcessor(final ProcessorData processorData) {
+    // Create ephemeral path in zookeeper corresponding to this processor.
+    ephemeralPath = zkClient.createEphemeralSequential(keyBuilder.getProcessorsPath() + "/", processorData.toString());
+    LOG.info("Created ephemeral path: {} for processor: {} in zookeeper.", ephemeralPath, processorData);
+    ProcessorNode processorNode = new ProcessorNode(processorData, ephemeralPath);
+    String processorId = processorData.getProcessorId();
+    // Determine if there are duplicate processors with this.processorId after registration.
+    if (getProcessorCount(processorId) > 1 && shouldStopProcessor(processorNode)) {
+      LOG.info("Processor: {} is duplicate. Deleting zookeeper node at path: {}.", processorData, ephemeralPath);
+      zkClient.delete(ephemeralPath);
+      throw new SamzaException(String.format("Processor: %s is duplicate in the group. Registration failed.", processorId));
     }
+    return ephemeralPath;
+  }
+
+  /**
+   * Determines if a processor should be stopped.
+   * @param processor to check for stop condition in processors group.
+   * @return true if processor should be stopped. false otherwise.
+   */
+  private boolean shouldStopProcessor(final ProcessorNode processor) {
+    /**
+     * If there are multiple processors registered with same processorId,
+     * allow only the processor with lexicographically smallest zookeeperPath to live
+     * and all the remaining processors should be stopped.
+     *
+     * Two processors will not have smallest zookeeperPath because of sequentialId guarantees
+     * of zookeeper for ephemeral nodes.
+     */
+    List<ProcessorNode> processorNodes = getAllProcessorNodes();
+    LOG.debug("All existing processor nodes in zookeeper: {}.", processorNodes);
+    TreeSet<String> sortedProcessorPaths = processorNodes.stream()
+                                                         .filter(processorNode -> processorNode.getProcessorData().getProcessorId().equals(processor.getProcessorData().getProcessorId()))
+                                                         .map(ProcessorNode::getEphemeralPath)
+                                                         .collect(Collectors.toCollection(TreeSet::new));
+    return !sortedProcessorPaths.first().equals(processor.getEphemeralPath());
+  }
+
+  /**
+   * Find the number of processors having the provided processorId.
+   * @param processorId to use for equality check.
+   * @return the count of matched processors.
+   */
+  private long getProcessorCount(final String processorId) {
```

Comment:
Can you modify this method to return the list of duplicate processors, if any, instead of the count? 
In Line 140, you are reading invoking `getAllProcessorNodes` twice - once in getProcessorsCount and shouldStopProcessor. Each call to `getAllProcessorNodes` will read the entire list and its data. It seems inefficient and also, potentially, you can get different lists.